### PR TITLE
feat(GITOPSRVCE-751): Add alert rule/test for Deployment SLO

### PR DESCRIPTION
**Pull request by jparsai**
## Description
This PR is to add alert rules and their tests for Deployments SLOs added by GitOps Service team.

## Jira Ticket
https://issues.redhat.com/browse/GITOPSRVCE-751

### CODE REVIEW

File: rhobs/alerting/data_plane/prometheus.gitops_deployment_alerts.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    tenant: rhtap
+spec:
+  groups:
+  - name: gitops_deployment_alerts
+    interval: 1m
+    rules:
+    - alert: GitOpsDeploymentHealthErrors
+      expr: |
+        (
+          sum(kube_deployment_status_replicas_ready{namespace="gitops-service-argocd"})
```
Comment (yftacherzog) [inline rhobs/alerting/data_plane/prometheus.gitops_deployment_alerts.yaml:15]: Do we need to `sum` here?

File: rhobs/alerting/data_plane/prometheus.gitops_routes_alerts.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    tenant: rhtap
+spec:
+  groups:
+  - name: gitops_routess_alerts
+    interval: 1m
+    rules:
+    - alert: GitOpsRoutesHealthErrors
+      expr: |
+        (
+          sum(openshift_route_status{ status="True", namespace="gitops-service-argocd"})
```
Comment (yftacherzog) [inline rhobs/alerting/data_plane/prometheus.gitops_routes_alerts.yaml:15]: What's the reason for using `sum` here?

File: rhobs/alerting/data_plane/prometheus.gitops_routes_alerts.yaml [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+metadata:
+  name: rhtap-gitops-routes-alerting
+  labels:
+    tenant: rhtap
+spec:
+  groups:
+  - name: gitops_routess_alerts
+    interval: 1m
+    rules:
+    - alert: GitOpsRoutesHealthErrors
```
Comment (yftacherzog) [inline rhobs/alerting/data_plane/prometheus.gitops_routes_alerts.yaml:12]: I don't see this metric on the tests.

**Review by yftacherzog — CHANGES_REQUESTED** [review]
Can we have separate PR for the different alerts?

File: rhobs/alerting/data_plane/prometheus.gitops_deployment_alerts.yaml (hunk 1 above)
Comment (jparsai) [reply rhobs/alerting/data_plane/prometheus.gitops_deployment_alerts.yaml:15]: Here we get series for multiple deployments, to handle that I used `SUM()`.

File: rhobs/alerting/data_plane/prometheus.gitops_routes_alerts.yaml (hunk 2 above)
Comment (jparsai) [reply rhobs/alerting/data_plane/prometheus.gitops_routes_alerts.yaml:15]: Here we get series for multiple routes, to handle that I used `SUM()`.

Comment (jparsai) [reply rhobs/alerting/data_plane/prometheus.gitops_routes_alerts.yaml:15]: Yes, as of now there is only one route created in `gitops-service-argocd` namespace, so there is only one series.

File: rhobs/alerting/data_plane/prometheus.gitops_deployment_alerts.yaml (hunk 1 above)
Comment (jparsai) [reply rhobs/alerting/data_plane/prometheus.gitops_deployment_alerts.yaml:15]: It is supposed to show 4 rows.

image

Comment (yftacherzog) [reply rhobs/alerting/data_plane/prometheus.gitops_deployment_alerts.yaml:15]: What cluster is this taken from?

File: rhobs/alerting/data_plane/prometheus.gitops_deployment_alerts.yaml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        (
+          sum(kube_deployment_spec_replicas{namespace="gitops-service-argocd"})
+          by (source_cluster)
+        ) < 0.95
+      for: 5m
+      labels:
+        severity: warning
+        slo: true
+      annotations:
+        summary: >-
```
Comment (yftacherzog) [inline rhobs/alerting/data_plane/prometheus.gitops_deployment_alerts.yaml:28]: I suggest adding the source_cluster to the summary/description

File: rhobs/alerting/data_plane/prometheus.gitops_deployment_alerts.yaml (hunk 1 above)
Comment (yftacherzog) [inline rhobs/alerting/data_plane/prometheus.gitops_deployment_alerts.yaml:15]: Does the `kube_deployment_status_replicas_ready` metric gives the number of replicas in ready state or it gives `1` when it's ready?
What does it return when no replicas are in ready state? 0? no data?

**Review by yftacherzog — COMMENTED** [review]
@jparsai PR looks good. I'm missing some understand of the metrics behavior to understand how the query looks like when deployments are unhealthy. See my comment on that.

File: rhobs/alerting/data_plane/prometheus.gitops_deployment_alerts.yaml (hunk 1 above)
Comment (jparsai) [reply rhobs/alerting/data_plane/prometheus.gitops_deployment_alerts.yaml:15]: It is taken from RHTAP Multi Tenant cluster.

Comment (jparsai) [reply rhobs/alerting/data_plane/prometheus.gitops_deployment_alerts.yaml:15]: Row will be there but `value` will be 0, PTAL at this screenshot where 1st deployment is not in healthy.

image

File: rhobs/alerting/data_plane/prometheus.gitops_deployment_alerts.yaml [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        )
+        /
+        (
+          sum(kube_deployment_spec_replicas{namespace="gitops-service-argocd"})
+          by (source_cluster)
+        ) < 0.95
+      for: 5m
+      labels:
+        severity: warning
+        slo: true
```
Comment (amisstea) [inline rhobs/alerting/data_plane/prometheus.gitops_deployment_alerts.yaml:26]:
```suggestion
        slo: "true"
```

We recently discovered a bug with boolean label values (see #138). Please update all the `slo` labels across your PRs. Sorry for the inconvenience!

**Review by yftacherzog — APPROVED** [review]
lgtm. Just see to @amisstea's comment and we can merge it.

File: rhobs/alerting/data_plane/prometheus.gitops_deployment_alerts.yaml (hunk 5 above)
Comment (jparsai) [reply rhobs/alerting/data_plane/prometheus.gitops_deployment_alerts.yaml]: NP, I updated it, PTAL.

File: rhobs/alerting/data_plane/prometheus.gitops_deployment_alerts.yaml (hunk 4 above)
Comment (jparsai) [reply rhobs/alerting/data_plane/prometheus.gitops_deployment_alerts.yaml:28]: Added it, PTAL.

**Review by yftacherzog — APPROVED** [review]
lgtm

**Review by amisstea — APPROVED** [review]

**Pull request merged**